content-s3: add toml config hook

garrettbslone opened the pull request:
This pr is a work in progress to allow the influential accessor information for the content-s3 module to be accessed from toml config files. There should be and initial conifg file passed into `flux start` with the `-o,--config-path=dir` option as well as an s3 credential file with the s3 access keys. Here is a example set up:

/home/user/config/config.toml:
```
[content-s3]
credential-file = "/home/user/.s3-creds"
uri = "http://my-s3-endpoint:9000"
bucket = "my-bucket"
virtual-host-style = false
```
/home/user/.s3-creds:
```
access-key-id = "access-key"
secret-access-key = "secret-key"
```

Note: the current sharness test will not work yet.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
+    if (!(cpy = strdup (uri)))
+        goto error;
+
+    if (yuarel_parse (&yuri, (char *) cpy) < 0) {
```
cast is unnecessary here

garlick replied:
`yuarel_parse()` doesn't set errno so you'll have to set it manually, o/w the RPC may succeed when it should fail.

garlick replied:
In latest update, the cast is gone but errno is still not set on `yuarel_parse()` failure.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
+        goto error;
+    }
+
+    ctx->cfg->hostname = yuri.host;
+    if (!(strncat (ctx->cfg->hostname, ":", sizeof(ctx->cfg->hostname) + 1))) {
+        snprintf(errbuff, eb_size, "failed to form hostname");
+        goto error;
+    }
```
Since you'll eventually need to free 'cpy', and yuri.host points within 'cpy', you probably want to make a whole new string for the host[:port].  It's also iffy to append to the yuri.host since for that to work properly you have to assume there is space, and that kind of violates libyuarel's interface abstraction.

A helper function might be nice here, e.g. something like (untested)
```c
static char *hostport (const char *host, const char *port)
{
    char *s;
    if (asprintf (&s, "%s%s%s", host, port ? ":" : "", port ? port : "") < 0)
        return NULL;
    return s;
}

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
+        goto error;
+    }
+
+    if (!(cpy = strdup (bucket)))
+        goto error;
```
extra copy of bucket which is then leaked?

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
+    if (!strncmp (yuri.scheme, "https", 5))
+        ctx->is_secure = 1;
+
+    if (!(tbl = tomltk_parse_file (cred_file, &toml_error))) {
+        snprintf(errbuff, eb_size, "failed to parse credentials file");
+        goto error;
+    }
```
I'd suggest moving the cred file parsing to a separate function.

Also if the parse fails, get the parse error messages that reference the line number etc from the toml_error and put into errbuf.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
     struct s3_config *cfg;
     flux_t *h;
     const char *hashfun;
+    int is_secure;
+    int is_virtual_host;
```
Shouldn't these be in the `struct s3_config` instead?

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
+    char errbuf[256];
+    const char *errstr = NULL;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        goto error;
```
You won't need to alloc a whole new ctx if those two values were in the cfg struct.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -11,22 +11,196 @@
+    if (flux_conf_reload_decode (msg, &conf) < 0)
+        goto error;
+
+    flux_log(h, LOG_WARNING, "config relod detectedin content-s3: changes will not take effect until next flux restart");
```
spelling errors, and this notification could be moved to after `parse_config()` succeeds.

Since the logs include the module name you might just shorten to something like:
> config-reload: changes will not take effect until flux is restarted

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -50,7 +48,61 @@ static void content_s3_destroy (struct content_s3 *ctx)
+
+    cfg->secret_key = secret_key;
+    cfg->access_key = access_key;
+
```
You're missing a `toml_free (tbl)` .
Also if access key fetch fails, need to free the secret key.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -79,67 +127,35 @@ static int parse_config (struct content_s3 *ctx,
+    
     sprintf(tmp, "%d", yuri.port);
-    if (!(strncat (ctx->cfg->hostname, tmp, sizeof(ctx->cfg->hostname) + sizeof(yuri.port)))) {
+    if (!(cfg->hostname = hostport (yuri.host, tmp))) {
```
Could just make this function take an integer port.  (Sorry when I suggested it earlier I was thinking the yuri.port was a string).  Looks like yuarel sets the port to zero if it's not set in the input, so hostport would just test for 0 instead of NULL for omitting :port from the output string.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -59,14 +111,10 @@ static int parse_config (struct content_s3 *ctx,
     const char *uri = NULL;
     const char *bucket = NULL;
     const char *cred_file = NULL;
+    int is_virtual_host = 0;
```
I think this function would be a little clearer if it just allocated and returned a `struct s3_config *`.  Then you wouldn't need all the tmp variables.

garlick replied:
In current update, `parse_config()` still updates a config, doesn't allocate it.  Any reason this was not implemented?

garrettbslone commented:
rebased to get the readthedocs changes and re-pushed.

garlick commented:
Thanks for addressing those things!  Getting there!

The PRs need to be refactored. It's OK to have incremental development in a PR while it's in progress but it complicates review if those commits are each doing a bunch of different things spread all over the code.

So I'll make some summary comments here instead of trying to make them in incremental diffs:
* some `struct s3_config *` structure members point to allocated strings that are leaked when `ctx->cfg` is freed in `content_s3_destroy()`
* same in `parse_config()` error path.  Suggest you make a destructor for it and call from those two places.
* and in that destructor, save/restore errno like in `content_s3_destroy()`
* `parse_config()` leaks "cpy"
* `parse_credentials()` potentially clobbers errno on the `goto error` paths with `free (secret_key)` and `toml_free()`.  Need errno save/restore around those as well.
* `config_reload_cb()` leaks cfg.  Just free it right after `parse_config()` since it's not used

garrettbslone commented:
Since the last round of review I added a `s3_config_destroy` function to free up all of the allocated memory from the s3_config object, fixed the leak of `cpy` in `parse_config`, fixed the leak of `cfg` in `config_reload_cb`, saved errno where it was being clobbered, and squashed commits down.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -27,6 +33,216 @@ struct content_s3 {
+        goto error;
+    }
+
+    if (!(raw = toml_raw_in (tbl, "secret-access-key"))) {
```
Need to set errno here.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -27,6 +33,216 @@ struct content_s3 {
+        goto error;
+    }
+    
+    if (!(cfg->hostname = hostport (yuri.host, yuri.port))) {
+        free (cpy);
+        snprintf(errbuff, eb_size, "failed to form hostname");
+        errno = EINVAL;
+        goto error;
+    }
```
`hostport()` already sets errno to ENOMEM in both failure cases.  Either save/restore across copy or just set to ENOMEM instead of EINVAL.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -27,6 +33,216 @@ struct content_s3 {
+    }
+
+    if (!(cfg->bucket = strdup (bucket))) {
+        free (cpy);
```
save/restore errno across free()

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -27,6 +33,216 @@ struct content_s3 {
+    return cfg;
+
+error:
+    s3_config_destroy (cfg);
```
you could also just add free(cpy) here (with save/restore of errno), and drop it from the other places you are freeing it.

garlick commented on src/modules/content-s3/s3.c:
```diff
@@ -81,6 +81,11 @@ int s3_init (struct s3_config *cfg, const char **errstr)
 {
     S3Status status = S3_initialize ("s3", S3_INIT_ALL, cfg->hostname);
 
+    if (cfg->is_virtual_host > 0)
```
Drop > 0 here and the next one.  The test should be "is nonzero" and that's the same as "true" in C.

garlick commented:
Made a few more review comments.  Please fix those and fix the conflict, and then I think we can remove WIP and give it one last review.

garrettbslone commented:
Can one of the reviewers re-run thee address-sanitizer check for me please? It seems to have timed out right at the end.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -27,6 +33,217 @@ struct content_s3 {
+
+    if (flux_conf_unpack (conf,
+                          &error,
+                          "{s:{s:s, s:s, s:s, s?:b} !}",
```
Was playing around with some tests and I happened to notice that the `!` ought to be inside inner braces.  It's supposed to catch keys in the `[content-s3]` table that are unknown.

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -117,7 +117,7 @@ static int parse_credentials (struct s3_config *cfg,
 }
 
-static char *hostport (const char *host, const char *port)
+static char *hostport (const char *host, int port)
```
belongs in the other commit

garlick commented on src/modules/content-s3/content-s3.c:
```diff
@@ -141,7 +141,6 @@ static struct s3_config *parse_config (const flux_conf_t *conf,
     const char *bucket = NULL;
     const char *cred_file = NULL;
     int is_virtual_host = 0;
-    char tmp[6];
```
belongs in the other commit

garlick commented on t/t0024-content-s3.t:
```diff
@@ -139,6 +163,17 @@ test_expect_success HAVE_JQ 'kvs-checkpoint.get foo returns same value' '
         test_cmp value2.exp value2.out
 '
 
+test_expect_success 'changing config file does not kill running backing store' '
+    err=0 && 
+    export tmp=$FLUX_CONF_DIR && export S3_CONF_DIR=/dev/null && \
+    flux config reload && \
+	for size in $SIZES; do \
+		if ! check_blob $size; then err=$(($err+1)); fi; \
+	done && \
+    export FLUX_CONF_DIR=$tmp && \
+	test $err -eq 0
+'
```
This test shouldn't need to set FLUX_CONF_DIR, and S3_CONF_DIR isn't used any more.  Loading blobs is one way to check that the module didn't die, but maybe a better test would be that your message about the config reload not taking effect was emitted to the logs, e.g.

A suggested replacement test might be:
```sh
test_expect_success 'config: reload with good config logs notice of no effect' '
        flux config reload 2>/dev/null &&
        flux dmesg | grep content-s3 | tail -1 >reload.log &&
        grep "changes will not take effect until next flux restart" reload.log
'
```

garlick commented on t/t0024-content-s3.t:
```diff
@@ -139,6 +163,17 @@ test_expect_success HAVE_JQ 'kvs-checkpoint.get foo returns same value' '
+    export FLUX_CONF_DIR=$tmp && \
+	test $err -eq 0
+'
+
```
It may be nice to add a few tests for config failures here.  They should be pretty easy to check since they will just cause the `flux config reload` command to fail, but not the module.

Here is a set of tests I was playing with to verify your error handling code (which all checked out fine except the one case with `!` noted above), for your consideration:
```sh
##
# Config file reload tests
##

test_expect_success 'config: save good config' '
        mv content-s3.toml content-s3.save
'

test_expect_success 'config: reload with extra content-s3 key fails' '
        cp content-s3.save content-s3.toml &&
        echo "extrakey = 42" >>content-s3.toml &&
        test_must_fail flux config reload
'

test_expect_success 'config: reload with malformed uri fails' '
        cp content-s3.save content-s3.toml &&
        sed -i -e "s/uri =.*$/uri = \"baduri\"/" \
                content-s3.toml &&
        test_must_fail flux config reload
'

test_expect_success 'config: reload with bad credential file path fails' '
        cp content-s3.save content-s3.toml &&
        sed -i -e "s/credential-file =.*$/credential-file = \"noexist\"/" \
                content-s3.toml &&
        test_must_fail flux config reload
'

test_expect_success 'config: restore good config' '
        mv -f content-s3.save content-s3.toml &&
        flux config reload
'
```

garlick commented on t/t0024-content-s3.t:
```diff
@@ -139,6 +163,73 @@ test_expect_success HAVE_JQ 'kvs-checkpoint.get foo returns same value' '
+	test_must_fail flux config reload
+'
+
+test_expect_success 'config: reload with extra s3 key fails' '
```
There is no check in the cred file parsing for an extra key.  This test is failing because the credential file path is still messed up from the previous test.  The test should probably be removed.

garlick commented on t/t0024-content-s3.t:
```diff
@@ -139,6 +163,73 @@ test_expect_success HAVE_JQ 'kvs-checkpoint.get foo returns same value' '
+
+test_expect_success 'save good config' '
+	mv content-s3.toml content-s3.save &&
+	mv creds/creds.toml creds/creds.save
```
If you move the creds file out of the way then the expected failures in subsequent tests could be masked by the missing cred file.  Suggest you copy not move it.

garlick commented on t/t0024-content-s3.t:
```diff
@@ -139,6 +163,73 @@ test_expect_success HAVE_JQ 'kvs-checkpoint.get foo returns same value' '
+'
+
+test_expect_success 'config: load module' '
+	flux module load content-s3 testing
```
Oopps the "testing" option is a cut-and-paste-o from the content-files test and has no meaning here.  I see it in a couple other places too, please remove.

garlick commented:
Restarted centos 7 builder that failed here
```
ERROR: test_terminus
====================

../../../config/tap-driver.sh: line 679:  2318 Segmentation fault      (core dumped) "$@"
ERROR: test_terminus.t - missing test plan
ERROR: test_terminus.t - exited with status 139 (terminated by signal 11?)
```
